README.md
# GitCracken

GitKraken utils for non-commercial use

Author: KillWolfVlad at [GitKraken-AUR](https://github.com/KillWolfVlad/GitKraken-AUR)

Working on `GNU/Linux` (without `snap`), `Windows` and `macOS`!

> WARNING! On `macOS` you should patch `GitKraken` only after first launch and full program closing!
>
> WARNING! Do not install `GitKraken` from `snap` on `Ubuntu`, patcher wont be able to work. Use common installation.

## Requirements

- `Node.js` v12 LTS or later
- `Yarn`
- `GitKraken v7.0+ and v8.0+` (latest tested v8.0.1)

NOTE: GitKraken v6.x are no more supported. Patches are still available on release page.

## Usage

- `yarn install`
- `yarn build`
- `node dist/bin/gitcracken.js --help` for more usage information

### Patcher

```bash
$ yarn run gitcracken patcher [options] [actions...]
```

`actions` - array of values (any order, any count)

> If `actions` is empty, will be used `auto` mode (ideal for beginners)

| Action   | Description                                 |
|----------|---------------------------------------------|
| `backup` | Backup `app.asar` file                      |
| `unpack` | Unpack `app.asar` file into `app` directory |
| `patch`  | Patch `app` directory                       |
| `pack`   | Pack `app` directory to `app.asar` file     |
| `remove` | Remove `app` directory                      |

| Option            | Description (if not defined, will be used `auto` value) |
|-------------------|---------------------------------------------------------|
| `-a`, `--asar`    | Path to `app.asar` file                                 |
| `-d`, `--dir`     | Path to `app` directory                                 |
| `-f`, `--feature` | Patcher feature (from [patches](patches) dir)           |

> You can invoke `-f`, `--feature` several times to apply all patches!

#### Examples

`Auto` patch installed `GitKraken` (maybe require `sudo` privileges on `GNU/Linux`)

```bash
$ yarn run gitcracken patcher
```

Use custom path to `app.asar`

```bash
$ yarn run gitcracken patcher --asar ~/Downloads/gitkraken/resources/app.asar
```

Use custom `actions` (`backup`, `unpack` and `patch`)

```bash
$ yarn run gitcracken patcher backup unpack patch
```

Mac from sudo

```bash
sh-3.2# cd /.../Projects\ From\ GIT/GitCracken/GitCracken/dist/bin
sh-3.2# yarn run gitcracken.js patcher --asar /Applications/GitKraken.app/Contents/Resources/app.asar
```

## Disable Automatic Update

Add this content to your `hosts` file:

```text
0.0.0.0 release.gitkraken.com
0.0.0.0 release.axocdn.com
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v0.11] - 2021-10-11
### Added
- Support GitKraken v8.0.1

### Fixed
- Fixed wrong package.json path when trying to identify GitKraken version number

## [v0.10] - 2021-09-09
### Added
- Support GitKraken v7.7.2
- New patch type "private_repo" (thanks to @maxzxwd)
- Added maintainers to info on logo

### Changed
- Splitted patches for version ***>=7.x && <=7.5.x*** and ***>=7.6.x*** (thanks to @ductran95)
- Use TypeScript's `resolveJsonModule` option to load packages.json instead using `fs-extra` Npm package

## [v0.9] - 2020-11-12
### Added
- Support GitKraken v7.4.0

## [v0.8] - 2020-02-26
### Added
- Support GitKraken v6.5.3
- Added short flags for command-line options

### Changed
- Set same default actions (`backup`, `unpack`, `patch`, `pack` and `remove`) for all platforms (Win, Linux, Mac)
- Other minor fixs

## [v0.7] - 2020-02-11
### Changed
- Synchronized all forks

## [v0.6.1] - 2019-07-30
### Changed
- Support GitKraken v6.0.0
  - Pro
  - Individual
  - Enterprise
    - Self-Hosted
    - Stand-Alone
  - Modes
    - Development
    - Staging
- Bump dependencies

## [v0.6.0] - 2018-01-08
### Added
- Support GitKraken Pro v6.0.0
- Bump dependencies

[Unreleased]: https://github.com/BoGnY/GitCracken/compare/v0.11...HEAD
[v0.11]: https://github.com/BoGnY/GitCracken/compare/v0.10...v0.11
[v0.10]: https://github.com/BoGnY/GitCracken/compare/v0.9...v0.10
[v0.9]: https://github.com/BoGnY/GitCracken/compare/v0.8...v0.9
[v0.8]: https://github.com/BoGnY/GitCracken/compare/v0.7...v0.8
[v0.7]: https://github.com/BoGnY/GitCracken/compare/v0.6.1...v0.7
[v0.6.1]: https://github.com/BoGnY/GitCracken/compare/v0.6.0...v0.6.1
[v0.6.0]: https://github.com/BoGnY/GitCracken/releases/tag/v0.6.0


src/logo.ts
import chalk from "chalk";
import * as figlet from "figlet";

import {packageJson} from "../global";

/**
 * Logo
 */
export class Logo {
  /**
   * Print logo
   * @param text Main text
   */
  public static print(text = "GitCracken"): void {
    console.log();
    console.log(figlet.textSync(text, "ANSI Shadow"));
    console.log(`${chalk.bold("• Description")}: ${packageJson.description}`);
    console.log(`${chalk.bold("• Version")}: ${packageJson.version}`);
    console.log(`${chalk.bold("• Author")}: ${packageJson.author}`);
    console.log(`${chalk.bold("• Maintainers")}: ${packageJson.maintainers.map((maintainer) => maintainer.name).join(", ")}`);
    console.log(`${chalk.bold("• License")}: ${packageJson.license}`);
    console.log(`${chalk.bold("• Home Page")}: ${packageJson.homepage}`);
    console.log();
  }
}


src/secFile.ts
import * as crypto from "crypto";

import * as fs from "fs-extra";

/**
 * Password protected file with JSON content
 * This class is fork of <https://gitlab.com/maxkorp/secure-storage> library
 */
export class SecFile {
  private readonly _fileName: string;
  private readonly _password: string;
  private readonly _algorithm: string;
  private _data?: object;

  /**
   * Password protected file with JSON content constructor
   * @param fileName File name
   * @param password Password
   * @param algorithm Algorithm
   */
  public constructor(
    fileName: string,
    password: string,
    algorithm: string = "aes256",
  ) {
    this._fileName = fileName;
    this._password = password;
    this._algorithm = algorithm;
  }

  /**
   * File name
   */
  public get fileName(): string {
    return this._fileName;
  }

  /**
   * Password
   */
  public get password(): string {
    return this._password;
  }

  /**
   * Algorithm
   */
  public get algorithm(): string {
    return this._algorithm;
  }

  /**
   * Data
   */
  public get data(): object | undefined {
    return this._data;
  }

  /**
   * Read data from disk
   * @throws Error
   */
  public read(): void {
    const decipher = crypto.createDecipher(this.algorithm, this.password); // TODO: deprecated!
    this._data = JSON.parse(
      Buffer.concat([
        decipher.update(fs.readFileSync(this.fileName)),
        decipher.final(),
      ]).toString("utf8"),
    );
  }
}


src/patcher.ts
import * as os from "os";
import * as path from "path";

import * as asar from "asar";
import * as diff from "diff";
import * as fs from "fs-extra";
import natsort from "natsort";

import {baseDir} from "../global";
import {CURRENT_PLATFORM, Platforms} from "./platform";

/**
 * Patcher options
 */
export interface IPatcherOptions {
  /**
   * app.asar file
   */
  readonly asar?: string;

  /**
   * app directory
   */
  readonly dir?: string;

  /**
   * Patcher features
   */
  readonly features: string[];
}

/**
 * Patcher
 */
export class Patcher {
  private static findAsarUnix(...files: string[]): string | undefined {
    return files.find((file) => fs.existsSync(file));
  }

  private static findAsarLinux(): string | undefined {
    return Patcher.findAsarUnix(
      "/opt/gitkraken/resources/app.asar", // Arch Linux
      "/usr/share/gitkraken/resources/app.asar", // deb & rpm
    );
  }

  private static findAsarWindows(): string | undefined {
    const gitkrakenLocal = path.join(os.homedir(), "AppData/Local/gitkraken");
    if (!fs.existsSync(gitkrakenLocal)) {
      return undefined;
    }
    const apps = fs
      .readdirSync(gitkrakenLocal)
      .filter((item) => item.match(/^app-\d+\.\d+\.\d+$/));
    let app = apps.sort(natsort({desc: true}))[0];
    if (!app) {
      return undefined;
    }
    app = path.join(gitkrakenLocal, app, "resources/app.asar");
    return fs.existsSync(app) ? app : undefined;
  }

  private static findAsarMacOS(): string | undefined {
    return Patcher.findAsarUnix(
      "/Applications/GitKraken.app/Contents/Resources/app.asar",
    );
  }

  private static findAsar(dir?: string): string | undefined {
    if (dir) {
      return path.normalize(dir) + ".asar";
    }
    switch (CURRENT_PLATFORM) {
      case Platforms.linux:
        return Patcher.findAsarLinux();
      case Platforms.windows:
        return Patcher.findAsarWindows();
      case Platforms.macOS:
        return Patcher.findAsarMacOS();
    }
  }

  private static findDir(asarFile: string): string {
    return path.join(
      path.dirname(asarFile),
      path.basename(asarFile, path.extname(asarFile)),
    );
  }

  private readonly _asar: string;
  private readonly _dir: string;
  private readonly _features: string[];

  /**
   * Patcher constructor
   * @param options Options
   */
  public constructor(options: IPatcherOptions) {
    const maybeAsar = options.asar || Patcher.findAsar(options.dir);
    if (!maybeAsar) {
      throw new Error("Can't find app.asar!");
    }
    this._asar = maybeAsar;
    this._dir = options.dir || Patcher.findDir(this.asar);
    this._features = options.features;
    if (!this.features.length) {
      throw new Error("Features is empty!");
    }
  }

  /**
   * app.asar file
   */
  public get asar(): string {
    return this._asar;
  }

  /**
   * app directory
   */
  public get dir(): string {
    return this._dir;
  }

  /**
   * Patcher features
   */
  public get features(): string[] {
    return this._features;
  }

  /**
   * Backup app.asar file
   * @throws Error
   */
  public backupAsar(): string {
    const backup = `${this.asar}.${new Date().getTime()}.backup`;
    fs.copySync(this.asar, backup);
    return backup;
  }

  /**
   * Unpack app.asar file into app directory
   * @throws Error
   */
  public unpackAsar(): void {
    asar.extractAll(this.asar, this.dir);
  }

  /**
   * Pack app directory to app.asar file
   * @throws Error
   */
  public packDirAsync(): Promise<void> {
    return asar.createPackage(this.dir, this.asar);
  }

  /**
   * Remove app directory
   * @throws Error
   */
  public removeDir(): void {
    fs.removeSync(this.dir);
  }

  /**
   * Patch app directory
   * @throws Error
   */
  public patchDir(): void {
    let patchesDir = "";
    const packageJson = require(path.join(this.dir, "package.json"));
    const version: string = packageJson.version;
    const versionNumbers = version.split(".").map((x) => Number(x));
    const majorVersion = versionNumbers[0];
    const minorVersion = versionNumbers[1];
    const patchVersion = versionNumbers[2];

    if (majorVersion === 7) { // Major version must be 7 or 8
      if (minorVersion <= 5) { // version >=7.x & <=7.5.x then use legacy patches
        patchesDir = "../patches/7.0";
      } else { // version >= 7.6.x then use newer patches
        patchesDir = "../patches/7.6";
      }
    } else if (majorVersion === 8) { // Major version must be 7 or 8
      patchesDir = "../patches/7.6";
    } else {
      throw new Error(`Can't patch ${version}... Only version 7.x and 8.0.x are supported!`);
    }

    for (const feature of this.features) {
      this.patchDirWithFeature(feature, patchesDir);
    }
  }

  private patchDirWithFeature(feature: string, patchesDir: string): void {
    const patches = diff.parsePatch(
      fs.readFileSync(
        path.join(baseDir, patchesDir, `${feature}.diff`),
        "utf8",
      ),
    );
    for (const patch of patches) {
      this.patchDirWithPatch(patch);
    }
  }

  private patchDirWithPatch(patch: diff.ParsedDiff): void {
    const sourceData = fs.readFileSync(
      path.join(this.dir, patch.oldFileName!),
      "utf8",
    );
    const sourcePatchedData = diff.applyPatch(sourceData, patch);
    if ((sourcePatchedData as any) === false) {
      throw new Error(`Can't patch ${patch.oldFileName}`);
    }
    if (patch.oldFileName !== patch.newFileName) {
      fs.unlinkSync(path.join(this.dir, patch.oldFileName!));
    }
    fs.writeFileSync(
      path.join(this.dir, patch.newFileName!),
      sourcePatchedData,
      "utf8",
    );
  }
}
